### PR TITLE
feat: create dut-factory method (RDT-587)

### LINKED ISSUE #231: Make Create `Dut` Instance Easier (RDT-554)
For now if we're going to create a `Dut` instance manually, we have to declare a lot of things manually

- `MessageQueue`, to collect the serial output
- `pexpect_logfile`, to init the pexpect process
- `listener`, to redirect the serial output to the pexpect_logfile
- ...

Set default values for the `Dut` class, or provide a factory function, to help users create `Dut` instance easier.

### PR DESCRIPTION
**Pull request by horw**
- created factory class for dut
closes #231

### CODE REVIEW

**Comment by horw** [comment]
Hi @hfudev, this is a very initial draft commit for the implementation of the Fabrica class. I have a few questions. I have only created the base Dut class for now, but I don't think it's enough for the users. Should we implement other Duts in the same manner? (for example, IdfDut, SerialDut, etc.) It is possible, but it might result in a lot of duplicated code.

**Comment by horw** [comment]
As I see this task, I will split it into the following tasks:
- [ ] Move function logic under fixtures to `dut_factory`.
- [ ] Test the moved solution.
- [ ] Create a `DutFactory` class using these `dut_factory` functions.

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+
+class DutFactory:
+    def __init__(self):
+        self.obj_stack = []
+
+    # The object has to be closed;
+    # otherwise, the process will still be here,
+    # but there are some problems with deletion, most likely because of the fork process.
+    # Call close instead delete.
+    def close(self):
```
Comment (horw) [inline pytest-embedded/pytest_embedded/dut_factory.py:552]: At first, I tried using __del__, but it seems there are some issues with the process. So changed to close

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+            'keyfile': None,
+            # common
+            'test_case_name': buff_parametrized_fixtures['test_case_name'],
+            '_meta': buff_parametrized_fixtures['_meta'],
+            # pre-initialized fixtures
+            'dut_index': dut_index_global,
+            '_pexpect_logfile': _pexpect_logfile,
+            'pexpect_proc': pexpect_proc,
+            'msg_queue': msg_queue,
+        }
```
Comment (horw) [inline pytest-embedded/pytest_embedded/dut_factory.py:630]: Need to have some discuss about default values and how users can pass new values. I used the `buff_parametrized_fixtures` variable to store the previous fixture result.

**Comment by horw** [comment]
It still seems rough, but it's starting to work. @hfudev, what do you think about doing it this way? Also, I commented on some issues here.

**Comment by horw** [comment]
@hfudev, Hello, I have cleaned the history and it seems like the test was passed.

**Comment by hfudev** [comment]
```python
# test_foo.py
from pytest_embedded.dut_factory import DutFactory


def test_foo():
    dut = DutFactory.create(app_path='components/console/test_apps/console', target='esp32c6')
    dut.run_all_single_board_cases()
```

unfortunately it's not working with error message

```
            _pexpect_logfile = os.path.join(
>               buff_parametrized_fixtures['_meta'].logdir, f'custom-dut-{dut_index_global}.txt'
            )
E           KeyError: '_meta'

```

Seems like the `set_buff_parametrized_fixtures` will be triggered in another fixture `parametrize_fixtures`. If I don't use any fixture, the `set_buff_parametrized_fixtures` will be skipped.

Besides, could you also add the `DutFactory` to `pytest_embedded/__init__.py` `__all__`?

**Comment by horw** [comment]
@hfudev 
Thanks for your review! Your testcase was very useful!

This commit contains the following changes:
- `parametrize_fixtures` changed to be autouse, which helps initialize `set_buff_parametrized_fixtures`.
- Added `unit_tester`.
- Added a test for `esp,idf`.
- Call `DutFactory.close()` after every test, not the session.

Do you have any other testcase I can try to implement?

Further steps:
- [ ] Add the `DutFactory` to `pytest_embedded/__init__.py ` `__all__`.
- [ ] Check `dut_factory` again, as there was some conflict related to the update in `plugin.py` when rebasing (I have just accepted my changes).
- [ ] Check for possible issues with `dut-{number}`.

File: pytest-embedded-idf/pytest_embedded_idf/serial.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/serial.py:1]: better to make this change in a separate PR to get release faster...

File: pytest-embedded-idf/pytest_embedded_idf/serial.py [hunk 3]
```diff
@@ -161,6 +161,8 @@ def flash(self, app: Optional[IdfApp] = None) -> None:
                 else:
                     _args.append(str(v))
 
+        _args.extend(['--baud', '1000000'])
```
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/serial.py:164]: let's keep the behavior the same as `idf.py flash`

```
❯ idf.py flash --help
Usage: idf.py flash [OPTIONS]

  Flash the project.

Options:
  -C, --project-dir PATH         Project directory.
  -D, --define-cache-entry TEXT  Create a cmake cache entry. This option can be used at most once either globally, or for one subcommand.
  -b, --baud INTEGER             Baud rate for flashing. It can imply monitor baud rate as well if it hasn't been defined locally. The
                                 default value can be set with the ESPBAUD environment variable. This option can be used at most once
                                 either globally, or for one subcommand.
```

i.e., if ESPBAUD set, use the set value. if not, use the suggested value 1000000

File: pytest-embedded-idf/pytest_embedded_idf/serial.py
Comment (horw) [reply pytest-embedded-idf/pytest_embedded_idf/serial.py:1]: Yes, I will do :+1:

File: pytest-embedded-idf/pytest_embedded_idf/serial.py (hunk 3 above)
Comment (horw) [reply pytest-embedded-idf/pytest_embedded_idf/serial.py]: Agree

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 4]
```diff
@@ -0,0 +1,682 @@
...
+        qemu_prog_path=None,
+        qemu_cli_args=None,
+        qemu_extra_args=None,
+        wokwi_cli_path=None,
+        wokwi_timeout=0,
+        skip_regenerate_image=None,
+        encrypt=None,
+        keyfile=None,
+    ):
+        layout = []
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/dut_factory.py:589]: for public APIs, let's write some documentations to tell users how to call it.

like 

```python
def create():
    """
    Create DUT object according to the arguments.

    Examples:
    >>> foo = DutFactory.create(embedded_services='idf,esp', app_path='path_to_hello_world')
    >>> foo.expect_exact('Hello world!')
    """
```

**Comment by horw** [comment]
> for example, if we support `--esptool-version` for `EspSerial, how many files shall we touch? (we won't support that, just an example) could you help document it somewhere in the comments as well?

Hello, let's take `wokwi_timeout` as an example. This is the output from `git grep`. As a result, you need to make changes in `dut_factory` and the `plugin`.
```
pytest-embedded-wokwi/pytest_embedded_wokwi/wokwi_cli.py:41:        wokwi_timeout: t.Optional[int] = None,
pytest-embedded-wokwi/pytest_embedded_wokwi/wokwi_cli.py:57:        if (wokwi_timeout is not None) and (wokwi_timeout > 0):
pytest-embedded-wokwi/pytest_embedded_wokwi/wokwi_cli.py:58:            cmd.extend(['--timeout', str(wokwi_timeout)])
pytest-embedded/pytest_embedded/dut_factory.py:140:    wokwi_timeout,
pytest-embedded/pytest_embedded/dut_factory.py:291:                    'wokwi_timeout': wokwi_timeout,
pytest-embedded/pytest_embedded/dut_factory.py:584:        wokwi_timeout=0,
pytest-embedded/pytest_embedded/dut_factory.py:638:                'wokwi_timeout': wokwi_timeout,
pytest-embedded/pytest_embedded/plugin.py:950:def wokwi_timeout(request: FixtureRequest) -> t.Optional[str]:
pytest-embedded/pytest_embedded/plugin.py:952:    return _request_param_or_config_option_or_default(request, 'wokwi_timeout', None)
pytest-embedded/pytest_embedded/plugin.py:1010:    wokwi_timeout,
```

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 5]
```diff
@@ -0,0 +1,682 @@
...
+            elif k == 'gdb':
+                kwargs[k] = gdb
+            elif k == 'qemu':
+                kwargs[k] = qemu
+            elif k == 'wokwi':
+                kwargs[k] = wokwi
+    return cls(**_drop_none_kwargs(kwargs), mixins=mixins)
+
+
+buff_parametrized_fixtures = {}
```
[the remark this answers is not in the document]
Comment (horw) [reply pytest-embedded/pytest_embedded/dut_factory.py:482]: For the first run, the buffer helps store some fixtures such as metadata, test case names, and services, etc., for reuse. It might be better to change the name to cache.

[the remark this answers is not in the document]
Comment (horw) [reply pytest-embedded/pytest_embedded/dut_factory.py:481]: Renamed to cache

File: pytest-embedded/pytest_embedded/dut_factory.py (hunk 4 above)
Comment (horw) [reply pytest-embedded/pytest_embedded/dut_factory.py:632]: Updated

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 6]
```diff
@@ -0,0 +1,725 @@
...
+        wokwi_cli_path=None,
+        wokwi_timeout=0,
+        skip_regenerate_image=None,
+        encrypt=None,
+        keyfile=None,
+    ):
+        """
+        Create a Device Under Test (DUT) object with customizable parameters.
+
+        Parameters:
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/dut_factory.py:591]: Here you're writing documentation in a non-standard way. This repo is using google docstring style. https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html#:~:text=def%20function_with_pep484_type_annotations(,False%20otherwise.%0A%0A%20%20%20%20%22%22%22

besides, it's much better to write the types in the function signatures than in the docstrings.

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 7]
```diff
@@ -0,0 +1,725 @@
...
+            elif k == 'gdb':
+                kwargs[k] = gdb
+            elif k == 'qemu':
+                kwargs[k] = qemu
+            elif k == 'wokwi':
+                kwargs[k] = wokwi
+    return cls(**_drop_none_kwargs(kwargs), mixins=mixins)
+
+
+parametrized_fixtures_cache = {}
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/dut_factory.py:481]: could you place all the global variables together at the top of this file? and make them uppercase, with some documentation explaining what are these global vars storing.

(increasing the readability)

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 8]
```diff
@@ -0,0 +1,725 @@
...
+        layout = []
+        try:
+            global parametrized_fixtures_cache
+            msg_queue = msg_queue_gn()
+            layout.append(msg_queue)
+
+            _pexpect_logfile = os.path.join(
+                parametrized_fixtures_cache['_meta'].logdir, f'custom-dut-{dut_global_index}.txt'
+            )
+            logging.warning('You can get your custom DUT log file at the following path: %s.', _pexpect_logfile)
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/dut_factory.py:641]:
```suggestion
            logging.debug('You can get your custom DUT log file at the following path: %s.', _pexpect_logfile)
```

either debug or info is fine, but not a warning.

File: pytest-embedded/pytest_embedded/dut_factory.py [hunk 9]
```diff
@@ -0,0 +1,725 @@
...
+                        _referrer[_i] = None
+            elif isinstance(_referrer, dict):
+                for key, value in _referrer.items():
+                    if value is obj:
+                        _referrer[key] = None
+        del obj
+
+
+class DutFactory:
+    obj_stack: t.ClassVar[t.List[t.List[t.Any]]] = []
```
Comment (hfudev) [inline pytest-embedded/pytest_embedded/dut_factory.py:531]:
```suggestion
    # stores the objects that required by each dut
    # [
    #   [app, serial, ..., dut]  # dut-0
    #   [app, serial, ..., dut]  # dut-1
    #   ...
    # ]
    obj_stack: t.ClassVar[t.List[t.List[t.Any]]] = []
```

**Review by hfudev — CHANGES_REQUESTED** [review]
@horw Thanks for this great PR. Only some minor comments regarding to the code conventions. Let's merge it today!
